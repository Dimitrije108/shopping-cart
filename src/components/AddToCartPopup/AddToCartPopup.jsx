import { useEffect, useRef } from "react";
import { Link } from "react-router-dom";
import HoverArrows from "../HoverArrows/HoverArrows";
import styles from "./AddToCartPopup.module.css";

export default function AddToCartPopup({ 
	id,
	quantity, 
	name,
	removeTimer,
	resetTimer,
}) {
	return (
		<Link 
			to="/shopping-cart" 
			className={styles.popupCont}
			onMouseEnter={() => removeTimer(id)}
			onMouseLeave={() => resetTimer(id)}
		>
			<div>
				<p>
					{quantity}x <span className={styles.name}>{name}</span>
				</p>
				<div className={styles.succesfullyAddedCont}>
					<p>succesfully added to cart!</p>
					<HoverArrows 
						showDetails={styles.showDetails}
						arrow={styles.arrow}
						width={12} 
						height={12}
					/>
				</div>
			</div>
		</Link>
	)
};
