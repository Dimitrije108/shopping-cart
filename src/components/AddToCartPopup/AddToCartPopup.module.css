.popupCont {
	max-width: 350px;
	min-height: 70px;
	padding: 1rem;
	box-sizing: border-box;

	font-family: DINNext;
	color: rgb(221, 221, 221);

	background-color: rgba(20, 20, 20, 0.1);
	outline: 1px solid rgb(104, 104, 104);
	backdrop-filter: blur(20px);

	position: fixed;
	bottom: 1rem;
	right: 1rem;
	z-index: 1;

	display: flex;
	align-items: center;

	animation: popup 1s cubic-bezier(0.25, 0.8, 0.25, 1);
}

.popupCont:hover {
	color: rgb(221, 221, 221);
	outline: 1px solid rgb(221, 221, 221);

	visibility: visible;
	cursor: pointer;
}

.name {
	font-family: DINNext;
	font-size: 1.4rem;
}

.succesfullyAddedCont {
	display: flex;
	justify-content: space-between;
}
/* 3 arrow animation indicating additional information is available */
.showDetails {
	padding-left: 0.5rem;
  z-index: 1;
  visibility: hidden;

  opacity: 0;
  transition: opacity 0.5s ease-in-out;
}

.popupCont:hover .showDetails {
	visibility: visible;
  opacity: 1;
  transition-delay: 0s;
}

.arrow:nth-child(1) {
  transition-delay: 0.1s;
}

.arrow:nth-child(2) {
  transition-delay: 0.2s;
}

.arrow:nth-child(3) {
  transition-delay: 0.3s;
}
/* popup animation */
@keyframes popup {
	0% {
    bottom: -300px;
		opacity: 0.1;
  }
	50% {
		opacity: 0.5;
	}
	100% {
    bottom: 1rem;
		opacity: 1;
  }
}
