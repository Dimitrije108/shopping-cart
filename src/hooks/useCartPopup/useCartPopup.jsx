import { useState, createContext, useContext } from 'react';

const PopupContext = createContext({
	popups: null,
	addPopup: () => {},
	removeTimer: () => {},
	resetTimer: () => {},
});
// Shopping cart functionality passed on to all components 
// via context provider
export function useAddToCartPopup() {
	const [popups, setPopups] = useState([]);
	// Store popup timers
	const [popupTimers, setPopupTimers] = useState(new Map());
	// Remove a popup and execute a remove popup animation
	const removePopup = (id) => {
		// change isExiting to true which should start the exit animation
		// set a timer for setPopups to execute when animation ends
		if (!popupTimers.has(id)) {
			setPopups((prevPopups) => prevPopups.filter((item) => item.id !== id));
		}

		setPopupTimers((prevTimers) => {
			const newTimers = new Map(prevTimers);
			if (newTimers.has(id)) {
				clearTimeout(newTimers.get(id));
				newTimers.delete(id);
			};
			return newTimers;
		}) 
	};
	// Add a new 'add to cart' popup
	const addPopup = (quantity, name) => {
		const id = crypto.randomUUID();
		const newPopup = { id, quantity, name };

		setPopups((prevPopups) => [...prevPopups, newPopup]);
		// Set an initial 3s timer to remove the popup
		const timer = setTimeout(() => {
			removePopup(id);
		}, 3000);

		setPopupTimers((prevTimers) => {
			const newTimers = new Map(prevTimers);
			newTimers.set(id, timer);
			return newTimers;
		})
	};
	// Remove the timer when user hovers the popup
	const removeTimer = (id) => {
		setPopupTimers((prevTimers) => {
			const newTimers = new Map(prevTimers);

			if (newTimers.has(id)) {
				clearTimeout(newTimers.get(id));
				newTimers.delete(id);
			}

			return newTimers;
		})
	};
	// Reset the timer upon "onMouseLeave" event
	const resetTimer = (id) => {
		// Set a new, shorter, timer to remove the popup
		const timer = setTimeout(() => {
			removePopup(id);
		}, 2000);
		// Reset the timer
		setPopupTimers((prevTimers) => {
			const newTimers = new Map(prevTimers);

			if (newTimers.has(id)) {
				clearTimeout(newTimers.get(id));
			}

			newTimers.set(id, timer);
			return newTimers;
		})
	};

	return { 
		popups, 
		addPopup, 
		removeTimer, 
		resetTimer,
	};
};

export const useCartPopup = () => useContext(PopupContext);

export default PopupContext;