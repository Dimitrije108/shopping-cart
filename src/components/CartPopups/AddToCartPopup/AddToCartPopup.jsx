import { Link } from "react-router-dom";
import PropTypes from 'prop-types';
import { useCartPopup } from "../../../hooks/useCartPopup/useCartPopup";
import HoverArrows from "../../HoverArrows/HoverArrows";
import styles from "./AddToCartPopup.module.css";

export default function AddToCartPopup({ 
	id,
	quantity = 1, 
	name = "Default",
	isExiting = false,
}) {
	const { removeTimer, resetTimer } = useCartPopup();
	
	return (
		<Link 
			to="/shopping-cart" 
			className={`${styles.popupCont} ${isExiting ? styles.exiting : ''}`}
			onMouseEnter={() => removeTimer(id)}
			onMouseLeave={() => resetTimer(id)}
		>
			<div>
				<p>
					{quantity}x <span className={styles.name}>{name}</span>
				</p>
				<div className={styles.succesfullyAddedCont}>
					<p>succesfully added to cart!</p>
					<HoverArrows 
						showDetails={styles.showDetails}
						arrow={styles.arrow}
						width={12} 
						height={12}
					/>
				</div>
			</div>
		</Link>
	)
};

AddToCartPopup.propTypes = {
	id: PropTypes.string.isRequired,
	quantity: PropTypes.number,
	name: PropTypes.string,
	isExiting: PropTypes.bool,
};
