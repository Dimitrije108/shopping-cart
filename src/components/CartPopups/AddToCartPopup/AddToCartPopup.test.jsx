import { describe, it, expect, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { MemoryRouter } from "react-router-dom";
import AddToCartPopup from "./AddToCartPopup";
import { useCartPopup } from "../../../hooks/useCartPopup/useCartPopup";
import styles from "./AddToCartPopup.module.css";

vi.mock("../../../hooks/useCartPopup/useCartPopup", () => ({
  useCartPopup: vi.fn(() => ({
		removeTimer: vi.fn(),
		resetTimer: vi.fn(),
	})),
}));

const mockPopup = {
	id: "6429291f021f17e13fbc1db9",
	quantity: 1,
	name: "Millennium Falcon",
	isExiting: false,
};

describe("Add To Cart Popup component", () => {
	it("add to cart popup component snapshot", () => {
		const popup = render(
			<MemoryRouter>
				<AddToCartPopup 
					id={mockPopup.id}
					quantity={mockPopup.quantity} 
					name={mockPopup.name}
					isExiting={mockPopup.isExiting}
				/>
			</MemoryRouter>
		)

		expect(popup).toMatchSnapshot();
	});

	it("renders add to cart popup elements correctly", () => {
		render(
			<MemoryRouter>
				<AddToCartPopup 
					id={mockPopup.id}
					quantity={mockPopup.quantity} 
					name={mockPopup.name}
					isExiting={mockPopup.isExiting}
				/>
			</MemoryRouter>
		)

		const link = screen.getByRole("link", { name: /succesfully added to cart!/i });
		const quantity = screen.getByText("1x");
		const name = screen.getByText("Millennium Falcon");
		const added = screen.getByText("succesfully added to cart!");
		const hoverArrows = screen.getAllByRole("img");

		expect(link).toBeInTheDocument();
		expect(quantity).toBeInTheDocument();
		expect(name).toBeInTheDocument();
		expect(added).toBeInTheDocument();
		expect(hoverArrows).toHaveLength(3);
	});
	
	it("removeTimer is called upon mouseenter", async () => {
		const user = userEvent.setup();
		
		const mockRemoveTimer = vi.fn();

		useCartPopup.mockReturnValue({
			removeTimer: mockRemoveTimer,
		});

		render(
			<MemoryRouter>
				<AddToCartPopup 
					id={mockPopup.id}
					quantity={mockPopup.quantity} 
					name={mockPopup.name}
					isExiting={mockPopup.isExiting}
				/>
			</MemoryRouter>
		)

		const link = screen.getByRole("link", { name: /succesfully added to cart!/i });

		await user.hover(link);

		expect(mockRemoveTimer).toHaveBeenCalledOnce();
	});

	it("resetTimer is called upon mouseleave", async () => {
		const user = userEvent.setup();
		
		const mockRemoveTimer = vi.fn();
		const mockResetTimer = vi.fn();

		useCartPopup.mockReturnValue({
			removeTimer: mockRemoveTimer,
			resetTimer: mockResetTimer,
		});

		render(
			<MemoryRouter>
				<AddToCartPopup 
					id={mockPopup.id}
					quantity={mockPopup.quantity} 
					name={mockPopup.name}
					isExiting={mockPopup.isExiting}
				/>
			</MemoryRouter>
		)

		const link = screen.getByRole("link", { name: /succesfully added to cart!/i });

		await user.hover(link);

		expect(mockRemoveTimer).toHaveBeenCalledOnce();

		await user.unhover(link);

		expect(mockResetTimer).toHaveBeenCalledOnce();
	});
	
	it("renders add to cart popup elements correctly", () => {
		render(
			<MemoryRouter>
				<AddToCartPopup 
					id={mockPopup.id}
					quantity={mockPopup.quantity} 
					name={mockPopup.name}
					isExiting={true}
				/>
			</MemoryRouter>
		)

		const link = screen.getByRole("link", { name: /succesfully added to cart!/i });

		expect(link).toBeInTheDocument();
		expect(link).toHaveClass(styles.exiting);
	});
});