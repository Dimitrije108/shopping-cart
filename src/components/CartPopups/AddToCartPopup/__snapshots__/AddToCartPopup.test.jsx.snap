// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Add To Cart Popup component > add to cart popup component snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <a
        class="_popupCont_32ea7d "
        data-discover="true"
        href="/shopping-cart"
      >
        <div>
          <p>
            1
            x 
            <span
              class="_name_32ea7d"
            >
              Millennium Falcon
            </span>
          </p>
          <div
            class="_succesfullyAddedCont_32ea7d"
          >
            <p>
              succesfully added to cart!
            </p>
            <div
              class="_showDetails_32ea7d"
            >
              <img
                class="_arrow_32ea7d"
                height="12"
                src="/src/assets/right-arrow.svg"
                width="12"
              />
              <img
                class="_arrow_32ea7d"
                height="12"
                src="/src/assets/right-arrow.svg"
                width="12"
              />
              <img
                class="_arrow_32ea7d"
                height="12"
                src="/src/assets/right-arrow.svg"
                width="12"
              />
            </div>
          </div>
        </div>
      </a>
    </div>
  </body>,
  "container": <div>
    <a
      class="_popupCont_32ea7d "
      data-discover="true"
      href="/shopping-cart"
    >
      <div>
        <p>
          1
          x 
          <span
            class="_name_32ea7d"
          >
            Millennium Falcon
          </span>
        </p>
        <div
          class="_succesfullyAddedCont_32ea7d"
        >
          <p>
            succesfully added to cart!
          </p>
          <div
            class="_showDetails_32ea7d"
          >
            <img
              class="_arrow_32ea7d"
              height="12"
              src="/src/assets/right-arrow.svg"
              width="12"
            />
            <img
              class="_arrow_32ea7d"
              height="12"
              src="/src/assets/right-arrow.svg"
              width="12"
            />
            <img
              class="_arrow_32ea7d"
              height="12"
              src="/src/assets/right-arrow.svg"
              width="12"
            />
          </div>
        </div>
      </div>
    </a>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
